### PR TITLE
Add pass_pipeline.md

### PR DESCRIPTION
**Pull request by linearhit**
add pass_pipeline.md

### CODE REVIEW

File: docs/developers/pass_pipeline.md [hunk 1]
```diff
@@ -0,0 +1,886 @@
...
+Before we start, let's recap the major problems we need to solve for the
+consideration of a dynamic shape compiler. Most of these aspects have been
+talked in [DISC : A Dynamic Shape Compiler for Machine Learning
+Workloads](https://arxiv.org/pdf/2103.05288v1.pdf).
+
+* IR definition in dynamic shape semantics
+
+We choose HLO as the tensor level "hub" IR interfacing with different frontend
+frameworks, since HLO has already been well established in the MHLO/XLA
+community and quite a lot of experience can be inherated. However, HLO was
```
Comment (yangjunpro) [inline docs/developers/pass_pipeline.md:26]: "inherated"=>"inherited"

File: docs/developers/pass_pipeline.md [hunk 2]
```diff
@@ -0,0 +1,886 @@
...
+semantics.
+
+* Shape calculation, placer, buffer management and runtime support
+
+One challenge of dynamic shape compiler is that, compiling is a static action,
+while we aim to handle dynamic behavior. In order to fully support dynamic
+shapes, the compiler must be able to do adaptive shape inference and generate
+code not only for data calculation, but also for shape calculation. The
+calculated shapes of each buffer will guide the buffer allocation and launch
+dimension at runtime. BladeDISC does bufferization during the conversion from
```
Comment (yangjunpro) [inline docs/developers/pass_pipeline.md:99]: "bufferization"==>"buffer assignment"?

File: docs/developers/pass_pipeline.md [hunk 3]
```diff
@@ -0,0 +1,886 @@
...
+* Performance
+
+The challenge of an AI compiler on performance is enlarged when the shapes are
+unknown or partially unknown. For the current version, BladeDISC makes an
+artificial seperation on compute-intensive part (gemm & convolution) and
+memory-intensive part (the rest calculations in general). For compute-intensive
+ops, different shapes may require different optimization to achieve the best
+performance. In order to balance the dynamism and performance, we implement an
+interface to choose the best kernel from a library according to different
+runtime shapes, and we do code generation for the memory intensive part. The
```
Comment (yangjunpro) [inline docs/developers/pass_pipeline.md:120]: Better to be consistent with either "memory-intensive" or "memory intensive" expression.

File: docs/developers/pass_pipeline.md [hunk 4]
```diff
@@ -0,0 +1,886 @@
...
+The code generation also faces the challenge of unknown shapes. A lot of regular
+optimization such as data vectorization, and codegen schedule selection will be
+more complicated when the sizes are unknown. For these aspects， BladeDISC will
+generate multiple versions of kernels at compile time, and generate the host
+side code to select a proper kernel to launch at runtime. We name this process
+as "speculation" as you will see in the consequent example.
+
+Moreover, without concrete shape value during compilation, we lose a wide
+variaty of optimization opportunities, from the graph level optimization such as
+algebriac simplifier, to the most regular CSE in the instruction level. This is
```
Comment (yangjunpro) [inline docs/developers/pass_pipeline.md:133]: I think CSE can also be applied at graph level.

File: docs/developers/pass_pipeline.md [hunk 5]
```diff
@@ -0,0 +1,886 @@
...
+dynamic shape cases. In principle, the static shape one should be in a higher
+priority for the consideration of performance.
+
+Besides of the standard tf to hlo pipeline of Mlir-HLO, and the DiscLowerTfPass
+that handles the custom-call ops special for BladeDISC such as RandomUniform and
+TopK, ReviseArgumentsForStaticRankPass is the only pass that needs your
+attention. This is due to that BladeDISC is a dynamic shape but "static rank"
+compiler considering of performance & complexity. Being a static rank compiler
+means that it is still possibly needed to compile more than once when the rank
+of the graph is changing, although this is rarely seen real AI workloads. Here
```
Comment (yangjunpro) [inline docs/developers/pass_pipeline.md:213]: "seen real AI workloads“==>"seen in real AI workloads".

File: docs/developers/pass_pipeline.md [hunk 6]
```diff
@@ -0,0 +1,886 @@
...
+and "static rank compiler".
+
+|               | static shape compiler | static rank compiler |
+| -------- | -------- | -------- |
+| requirements  | Any factors that decide the shape should be compile time const | Any factors that decide the rank should be compile time const |
+| inputs to a cluster are categorized into | (1) fixed shape inputs<br>(2) constant inputs | (1) fixed rank inputs<br>(2) fixed shape inputs<br>(3) constant inputs |
+| keys to a compilation cache, or, the infos must be logged during JIT clustering | shapes of (1) & values of (2) | ranks of (1) & shapes of (2) & values of (3) |
+
+If the value or shape of some inputs to a cluster decides the rank of some
+internal tensor, which is analyzed during clustering, the value or shape will to
```
Comment (yangjunpro) [inline docs/developers/pass_pipeline.md:224]: "will to be"==>"will be".

File: docs/developers/pass_pipeline.md [hunk 7]
```diff
@@ -0,0 +1,886 @@
...
+  %0 = tensor.cast %arg1 : tensor<10xf32> to tensor<?xf32>
+  %1 = "mhlo.add"(%arg0, %0) : (tensor<?xf32>, tensor<?xf32>) ->
+  tensor<?xf32> return %1 : tensor<?xf32>
+}
+```
+
+can be optimized into:
+
+```
+func @main(%arg0 : tensor<10xf32>, %arg1 : tensor<10xf32>) -> tensor<10xf32> {
```
Comment (yangjunpro) [inline docs/developers/pass_pipeline.md:266]: In my understanding, %arg0 can be optimized as tensor<10xf32> is due to the nature of mhlo.add operation. If my understanding is correct, is it safe to do this optimization based on the function body?

File: docs/developers/pass_pipeline.md [hunk 8]
```diff
@@ -0,0 +1,886 @@
...
+
+```python
+!sed -n '/After\ DiscShapeSimplifierPass/,/After\ Canonicalizer/p' pass_pipeline.log | sed '1d' | sed -n '/After\ DiscShapeSimplifierPass/,/After\ Canonicalizer/p'
+```
+
+# 3 Bufferize Passes
+
+BladeDISC does both host side and device side compilation. Thus the whole IR
+module will be lowered to a LLVM module eventually. The first step needed for
+this purpose is to go to the buffer world After we finish the high level
```
Comment (yangjunpro) [inline docs/developers/pass_pipeline.md:362]: "After we finish"==>"after finishing“

File: docs/developers/pass_pipeline.md [hunk 9]
```diff
@@ -0,0 +1,886 @@
...
+
+
+```python
+!sed -n '/After\ DiscAssignMemorySpacePass/,/After\ PromoteBuffersToStack/p' pass_pipeline.log
+```
+
+# 4 LHLO Graph Optimization Passes
+
+This phase contains some graph optimization passes that is better to be
+implemented on buffer level other than tensor level.
```
Comment (yangjunpro) [inline docs/developers/pass_pipeline.md:458]: "other than"=>”rather than“

File: docs/developers/pass_pipeline.md [hunk 10]
```diff
@@ -0,0 +1,886 @@
...
+
+## 4.1 Fusion Pass
+
+The most important graph optimization pass is the fusion pass, which provides a
+fusion strategy to seperate the lmhlo ops into a number of lmhlo.fusion, as the
+readers can see in the output IRs. An lmhlo.fusion will be further lowered into
```
Comment (yangjunpro) [inline docs/developers/pass_pipeline.md:464]: ”An"==>"A"？

File: docs/developers/pass_pipeline.md [hunk 11]
```diff
@@ -0,0 +1,886 @@
...
+
+The different "disc_vectorize_hint" attributes will guide the codegen passes to
+generate different versions of kernels.
+
+## 4.3 Memory Optimization Passes
+
+The other two passes are related with memory optimization, which can also be
+regarded as part of bufferization.
+
+The PromoteBuffersToStack pass, reusing the building block in MLIR repo, promote
```
Comment (yangjunpro) [inline docs/developers/pass_pipeline.md:537]: So "PromoteBuffersToStack" is a CPU-specific pass, right?

File: docs/developers/pass_pipeline.md [hunk 12]
```diff
@@ -0,0 +1,886 @@
...
+
+```python
+!sed -n '/After\ BufferDeallocation/,/After\ RalInjectExecutionContextPass/p' pass_pipeline.log
+```
+
+# 5 Runtime & Library Call Related Passes
+
+BladeDISC relies on Runtime Abstraction Layer (RAL) to manage stateful stuff and
+isolate the difference among different targeting environments (e.g. TensoroFlow,
+Pytorch, or even standalone execution). For the fully detail of RAL, please
```
Comment (yangjunpro) [inline docs/developers/pass_pipeline.md:557]: "fully"==>"full"..

File: docs/developers/pass_pipeline.md [hunk 13]
```diff
@@ -0,0 +1,894 @@
+# Abstract
+
+This tutorial will give a brief introduction for the compiler part of BladeDISC
+by walking through the backbone pass pipeline, hoping the readers can quickly
+build some intuitive understanding of the overall framework. We choose a small
```
Comment (linearhit) [inline docs/developers/pass_pipeline.md:5]: test robot

**Review by linearhit — COMMENTED** [review]
test robot

**Review by linearhit — COMMENTED** [review]
test bot

File: docs/developers/pass_pipeline.md [hunk 14]
```diff
@@ -0,0 +1,894 @@
...
+piece of TensorFlow subgraph as a sample that contains the major factors to be
+considered in the design, and will show you the brief lowering process and
+what does each backbone pass do to the IRs.
+
```
Comment (linearhit) [inline docs/developers/pass_pipeline.md:9]: test bot

File: docs/developers/pass_pipeline.md [hunk 15]
```diff
@@ -0,0 +1,894 @@
+# Abstract
+
+This tutorial will give a brief introduction for the compiler part of BladeDISC
+by walking through the backbone pass pipeline, hoping the readers can quickly
+build some intuitive understanding of the overall framework. We choose a small
+piece of TensorFlow subgraph as a sample that contains the major factors to be
+considered in the design, and will show you the brief lowering process and
```
Comment (linearhit) [inline docs/developers/pass_pipeline.md:7]: test bot

**Review by linearhit — COMMENTED** [review]
test bot

**Review by wyzero — APPROVED** [review]
LGTM

**Pull request merged**